File: firmware_v1.py
from dataclasses import dataclass
from enum import Enum, auto
import time
from collections import deque
from typing import Optional
import numpy as np

'''
MAKE SURE YOU INSTALL NUMPY
'''

'''
logic notes:
confidence level: 0..1 indicating how sure the system is that there is a kid
ts = timestamp
Optional means variable is allowed to have a real value or None
'''

# global states
class State(Enum):
    IDLE = auto()
    MONITOR = auto()
    SUSPECT = auto()
    CONFIRMED = auto()
    ALERTING = auto()

### NEED TO DEVELOP ALGOS FOR ALL VARIABLES IN DATACLASSES ###
# consult the sensor datasheets for outputs

# data class for thermal array
# 8x8 grid array
@dataclass
class ThermalReading:
    # use AMG88XX library to read data, calibrate, write to array
    temp_array = np.array([
        [],
        [],
        [],
        [],
        [],
        [],
        [],
        []
    ])
    ts: float = time.time()
    t_avg: float = np.mean(temp_array) # avg temp across array
    t_max: float = np.max(temp_array) # highest temp in array
    hotspot_area: float # 0..1, percent of image occupied by hottest region
    hotspot_count: float = 0
    for t in temp_array:
        if t >= t_max * 0.9: # tune this
            hotspot_count += 1.0
    hotspot_area = hotspot_count / 64.0
    blob_score: float    # 0..1, percent estimate of how human-like the heat blob is

# data class for CO2 sensor
@dataclass
class CO2Reading:
    ts: float = time.time()
    ppm: float = # ppm concentration
    ppm_rate_per_min: float # rate of rise/fall of ppm concentration

# data class for motion sensor
@dataclass
class MotionReading:
    ts: float = time.time()
    motion: bool # is there motion? T/F
    confidence: float = 1.0  # 0..1 if available
    # starts at 1.0 and changed later by logic
    # helps grade how meaningful a motion detected is

@dataclass
# OBD
class OBDReading:
    # how to read CAN -> UART -> C for Arduino
    # if the engine is on/off
    # how long has it been since front & rear doors were opened/closed
    # general internal cabin temperature

# clamping function, makes sure the logic stays between 0 and 1
def clamp(x, lo=0.0, hi=1.0):
    return max(lo, min(hi, x))

# hot car detector object
class ChildHotCarDetector:
    # tunables (start points, adjust with real data)
    HOT_CABIN_C = 35.0 # lower temp limit for what is considered a hot cabin
    HOTSPOT_HUMAN_MIN_C = 33.0 # lower temp limit for IR array
    CO2_HIGH_PPM = 1200.0 # lower limit for dangerous CO2 concentration
    CO2_RISE_PPM_PER_MIN = 100.0 # lower limit for rate of CO2 rise
    MOTION_RECENT_SEC = 20.0 # motion was last seen N seconds ago, if last motion time is less than this, it is still relevant

    SUSPECT_ENTER_CONF = 0.55 # minimum confidence level to enter suspect state
    CONFIRM_CONF = 0.78 # min conf level to confirm
    CLEAR_CONF = 0.35 # max conf level to clear state and return suspect -> monitor or alert -> idle

    SUSPECT_HOLD_SEC = 20.0 # how long conf level has to stay above suspect threshold before entering suspect state
    CONFIRM_HOLD_SEC = 30.0 # how long conf level has to stay above confirm threshold before confirm state
    CLEAR_HOLD_SEC = 60.0 # how long conf level has to stay below clear threshold before leaving activated state

    SENSOR_TIMEOUT_SEC = 5.0 # max allowed time between sensor readings before signaling potential sensor stale or failure
    # stale = stops reporting real time data even though appears to be working

    # weights (must sum <= ~1.2 then clamp)
    # how much each sensor contributes to overall confidence level, relativity matters
    WT = 0.50
    WC = 0.35
    WM = 0.25


    def __init__(self):
        self.state = State.IDLE # start with idle state
        self.state_enter_ts = time.time() # time that a state was entered

        # initialize sensor readings
        self.thermal: Optional[ThermalReading] = None
        self.co2: Optional[CO2Reading] = None
        self.motion: Optional[MotionReading] = None

        # initialize timestamps
        self.last_motion_ts: float = -1e9
        self.last_valid_thermal_ts: float = -1e9
        self.last_valid_co2_ts: float = -1e9
        self.last_valid_motion_ts: float = -1e9

        # initialize timestamp calculators
        self._conf_above_since: Optional[float] = None
        self._conf_below_since: Optional[float] = None

        # T/F is system in alert state?
        self.alerting = False

    def _heat_risk(self) -> float:
        if not self.thermal: # if there is no thermal reading stored, return 0 heat risk
            return 0.0
        # heat risk - weighted avg of t_avg + max
        t_avg = self.thermal.t_avg
        t_max = self.thermal.t_max
        # map 25C..45C to 0..1
        # change range -> change formulas
        # 25C = start of meaningful heat concern
        # 30C = starting reference point for when hotspots matter (moderate risk)
        # 45C = lower limit of extremely dangerous risk
        # divide by size of temp window to convert to dimensionless temp
        risk_avg = clamp((t_avg - 25.0) / 20.0) # dimensionless temp
        risk_max = clamp((t_max - 30.0) / 20.0)
        # weighted average of avg temp + max temp, can change weights if needed
        return clamp(0.6 * risk_avg + 0.4 * risk_max)

    def _occupant_score(self, now: float) -> float:
        # reset occupant score
        score = 0.0

        # thermal evidence
        if self.thermal:
            # reset thermal evidence score
            thermal_evidence = 0.0
            # add all indicators of hotspot to thermal evidence score, with weights
            if self.thermal.t_max >= self.HOTSPOT_HUMAN_MIN_C:
                thermal_evidence += 0.5
            thermal_evidence += 0.3 * clamp(self.thermal.hotspot_area)
            thermal_evidence += 0.2 * clamp(self.thermal.blob_score)
            # add thermal evidence score to total occupant score
            score += self.WT * clamp(thermal_evidence)

        # CO2 evidence
        if self.co2:
            # reset CO2 evidence score
            co2_evidence = 0.0
            # add all indicators of CO2 levels to CO2 score
            # weights do not add up to 1 because either can indicate risk
            if self.co2.ppm >= self.CO2_HIGH_PPM:
                co2_evidence += 0.7
            if self.co2.ppm_rate_per_min >= self.CO2_RISE_PPM_PER_MIN:
                co2_evidence += 0.6
            # add CO2 evidence score to total occupant score
            score += self.WC * clamp(co2_evidence)

        # motion evidence
        # T/F is there recent motion? compare most recent motion time to upper limit
        motion_recent = (now - self.last_motion_ts) <= self.MOTION_RECENT_SEC
        if motion_recent:
            # compute mconf level, currently can be only 0, 1, or direct confidence reading put into ingest_motion
            # check if mmWave module outputs detection confidence, signal strength, energy, or presence probability
            # if not, develop algo for detecting consecutive motion frames, reflection strength, small vs large motion, persistence of motion, filtering noise
            mconf = 1.0 if not self.motion else clamp(self.motion.confidence)
            score += self.WM * mconf

        # OBD evidence added to occupant score 

        return clamp(score)

    # T/F has the sensor reported data within the time limit for new readings?
    def _sensor_ok(self, last_ts: float, now: float) -> bool:
        return (now - last_ts) <= self.SENSOR_TIMEOUT_SEC

    # stores thermal reading and timestamp
    def ingest_thermal(self, r: ThermalReading):
        self.thermal = r
        self.last_valid_thermal_ts = r.ts

    # stores CO2 reading and timestamp
    def ingest_co2(self, r: CO2Reading):
        self.co2 = r
        self.last_valid_co2_ts = r.ts

    # stores motion reading and timestamp
    def ingest_motion(self, r: MotionReading):
        self.motion = r
        self.last_valid_motion_ts = r.ts
        if r.motion:
            self.last_motion_ts = r.ts

    def _transition(self, new_state: State, now: float):
        if new_state != self.state:
            # changes system state to current state
            self.state = new_state
            # records when new state started
            self.state_enter_ts = now
            # resets persistence timers
            self._conf_above_since = None
            self._conf_below_since = None

    # main method for process
    def update(self, now: Optional[float] = None):
        now = time.time() if now is None else now # find current timestamp

        # (optional) take count of working sensors
        thermal_ok = self._sensor_ok(self.last_valid_thermal_ts, now)
        co2_ok = self._sensor_ok(self.last_valid_co2_ts, now)
        motion_ok = self._sensor_ok(self.last_valid_motion_ts, now)

        # calculate total confidence score
        heat = self._heat_risk()
        occ = self._occupant_score(now)
        confidence = clamp(heat * occ)

        # THINK ABOUT THIS AND CHANGE HOW MUCH TO REDUCE CONF LEVEL BY
        # if too many sensors missing, reduce confidence
        ok_count = sum([thermal_ok, co2_ok, motion_ok])
        if ok_count <= 1:
            confidence *= 0.5

        # timers for persistence
        def held_above(thresh: float, hold_sec: float) -> bool:
            if confidence >= thresh:
                # if no previous high conf exists, set to now
                if self._conf_above_since is None:
                    self._conf_above_since = now
                # if previous high conf exists, calculate amount of time that high conf has been held
                # return true and transition to higher state
                return (now - self._conf_above_since) >= hold_sec
            else:
                # if conf below threshold, reset the timer
                self._conf_above_since = None
                return False

        def held_below(thresh: float, hold_sec: float) -> bool:
            if confidence <= thresh:
                if self._conf_below_since is None:
                    self._conf_below_since = now
                return (now - self._conf_below_since) >= hold_sec
            else:
                self._conf_below_since = None
                return False

        # state machine
        if self.state == State.IDLE:
            self.alerting = False
            if heat > 0.4:
                self._transition(State.MONITOR, now)

        elif self.state == State.MONITOR:
            self.alerting = False
            if held_above(self.SUSPECT_ENTER_CONF, self.SUSPECT_HOLD_SEC):
                self._transition(State.SUSPECT, now)
            elif heat < 0.2 and held_below(0.2, 10.0):
                self._transition(State.IDLE, now)

        elif self.state == State.SUSPECT:
            self.alerting = False
            if held_above(self.CONFIRM_CONF, self.CONFIRM_HOLD_SEC):
                self._transition(State.CONFIRMED, now)
            elif held_below(self.CLEAR_CONF, self.CLEAR_HOLD_SEC):
                self._transition(State.MONITOR, now)

        elif self.state == State.CONFIRMED:
            # immediate to alerting
            self.alerting = True
            self._transition(State.ALERTING, now)

        elif self.state == State.ALERTING:
            self.alerting = True
            # clear conditions
            if heat < 0.2 and held_below(self.CLEAR_CONF, self.CLEAR_HOLD_SEC):
                self.alerting = False
                self._transition(State.IDLE, now)

        return {
            "state": self.state.name,
            "heat_risk": heat,
            "occupant_score": occ,
            "confidence": confidence,
            "thermal_ok": thermal_ok,
            "co2_ok": co2_ok,
            "motion_ok": motion_ok,
            "alerting": self.alerting
        }


# example test with fake data
if __name__ == "__main__":
    d = ChildHotCarDetector()
    t0 = time.time()

    for i in range(0, 120):
        now = t0 + i

        # fake thermal ramp
        d.ingest_thermal(ThermalReading(
            ts=now, t_avg=30 + 0.05*i, t_max=33 + 0.08*i,
            hotspot_area=0.4, blob_score=0.6
        ))

        # fake CO2 rising
        d.ingest_co2(CO2Reading(ts=now, ppm=800 + 6*i, ppm_rate_per_min=120))

        # fake motion bursts
        d.ingest_motion(MotionReading(ts=now, motion=(i % 15 == 0), confidence=0.9))

        out = d.update(now)
        if i % 10 == 0:
            print(i, out)

File: firmware_v1.c
#include <Arduino.h>
#include <math.h>

/*
logic notes:
confidence level: 0..1 indicating how sure the system is that there is a kid
ts = timestamp
Optional means variable is allowed to have a real value or None
*/

// global states
typedef enum {
  STATE_IDLE = 0,
  STATE_MONITOR,
  STATE_SUSPECT,
  STATE_CONFIRMED,
  STATE_ALERTING
} State;

static const char* state_name(State s) {
  switch (s) {
    case STATE_IDLE: return "IDLE";
    case STATE_MONITOR: return "MONITOR";
    case STATE_SUSPECT: return "SUSPECT";
    case STATE_CONFIRMED: return "CONFIRMED";
    case STATE_ALERTING: return "ALERTING";
    default: return "UNKNOWN";
  }
}

### NEED TO DEVELOP ALGOS FOR ALL VARIABLES IN DATACLASSES ###
/* consult the sensor datasheets for outputs */

// data class for thermal array
// 8x8 grid array
typedef struct {
  // use AMG88XX library to read data, calibrate, write to array
  float temp_array[8][8];

  float ts;
  float t_avg; // avg temp across array
  float t_max; // highest temp in array
  float hotspot_area; // 0..1, percent of image occupied by hottest region
  float blob_score;   // 0..1, percent estimate of how human-like the heat blob is
} ThermalReading;

// data class for CO2 sensor
typedef struct {
  float ts;
  int th; // high level output time
  int to; // cycle output time (1004 ms +/- 5%)
  float ppm; // ppm concentration
  float ppm_rate_per_min; // rate of rise/fall of ppm concentration
} CO2Reading;

// data class for motion sensor
typedef struct {
  float ts;
  bool motion; // is there motion? T/F
  float confidence;  // 0..1 if available
  // starts at 1.0 and changed later by logic
  // helps grade how meaningful a motion detected is
} MotionReading;

typedef struct {
} OBDReading;

@dataclass
// OBD
/* class OBDReading:
    # how to read CAN -> UART -> C for Arduino
    # if the engine is on/off
    # how long has it been since front & rear doors were opened/closed
    # general internal cabin temperature
*/

// clamping function, makes sure the logic stays between 0 and 1
static float clampf(float x, float lo = 0.0f, float hi = 1.0f) {
  if (x < lo) return lo;
  if (x > hi) return hi;
  return x;
}

// hot car detector object
class ChildHotCarDetector {
public:
  // tunables (start points, adjust with real data)
  const float HOT_CABIN_C = 35.0f; // lower temp limit for what is considered a hot cabin
  const float HOTSPOT_HUMAN_MIN_C = 33.0f; // lower temp limit for IR array
  const float CO2_HIGH_PPM = 1200.0f; // lower limit for dangerous CO2 concentration
  const float CO2_RISE_PPM_PER_MIN = 100.0f; // lower limit for rate of CO2 rise
  const float MOTION_RECENT_SEC = 20.0f; // motion was last seen N seconds ago, if last motion time is less than this, it is still relevant

  const float SUSPECT_ENTER_CONF = 0.55f; // minimum confidence level to enter suspect state
  const float CONFIRM_CONF = 0.78f; // min conf level to confirm
  const float CLEAR_CONF = 0.35f; // max conf level to clear state and return suspect -> monitor or alert -> idle

  const float SUSPECT_HOLD_SEC = 20.0f; // how long conf level has to stay above suspect threshold before entering suspect state
  const float CONFIRM_HOLD_SEC = 30.0f; // how long conf level has to stay above confirm threshold before confirm state
  const float CLEAR_HOLD_SEC = 60.0f; // how long conf level has to stay below clear threshold before leaving activated state

  const float SENSOR_TIMEOUT_SEC = 5.0f; // max allowed time between sensor readings before signaling potential sensor stale or failure
  // stale = stops reporting real time data even though appears to be working

  // weights (must sum <= ~1.2 then clamp)
  // how much each sensor contributes to overall confidence level, relativity matters
  const float WT = 0.50f;
  const float WC = 0.35f;
  const float WM = 0.25f;

  ChildHotCarDetector() {
    this->state = STATE_IDLE; // start with idle state
    this->state_enter_ts = now_s(); // time that a state was entered

    // initialize sensor readings
    this->has_thermal = false;
    this->has_co2 = false;
    this->has_motion = false;

    // initialize timestamps
    this->last_motion_ts = -1e9f;
    this->last_valid_thermal_ts = -1e9f;
    this->last_valid_co2_ts = -1e9f;
    this->last_valid_motion_ts = -1e9f;

    // initialize timestamp calculators
    this->_conf_above_since_set = false;
    this->_conf_below_since_set = false;
    this->_conf_above_since = 0.0f;
    this->_conf_below_since = 0.0f;

    // T/F is system in alert state?
    this->alerting = false;
  }

  // stores thermal reading and timestamp
  void ingest_thermal(const ThermalReading &r) {
    this->thermal = r;
    this->has_thermal = true;
    this->last_valid_thermal_ts = r.ts;
  }

  // stores CO2 reading and timestamp
  void ingest_co2(const CO2Reading &r) {
    this->co2 = r;
    this->has_co2 = true;
    this->last_valid_co2_ts = r.ts;
  }

  // stores motion reading and timestamp
  void ingest_motion(const MotionReading &r) {
    this->motion = r;
    this->has_motion = true;
    this->last_valid_motion_ts = r.ts;
    if (r.motion) {
      this->last_motion_ts = r.ts;
    }
  }

  typedef struct {
    const char* state;
    float heat_risk;
    float occupant_score;
    float confidence;
    bool thermal_ok;
    bool co2_ok;
    bool motion_ok;
    bool alerting;
  } Output;

  // main method for process
  Output update(float now_in = NAN) {
    float now = isnan(now_in) ? now_s() : now_in; // find current timestamp

    // (optional) take count of working sensors
    bool thermal_ok = _sensor_ok(this->last_valid_thermal_ts, now);
    bool co2_ok = _sensor_ok(this->last_valid_co2_ts, now);
    bool motion_ok = _sensor_ok(this->last_valid_motion_ts, now);

    // calculate total confidence score
    float heat = _heat_risk();
    float occ = _occupant_score(now);
    float confidence = clampf(heat * occ);

    // THINK ABOUT THIS AND CHANGE HOW MUCH TO REDUCE CONF LEVEL BY
    // if too many sensors missing, reduce confidence
    int ok_count = (thermal_ok ? 1 : 0) + (co2_ok ? 1 : 0) + (motion_ok ? 1 : 0);
    if (ok_count <= 1) {
      confidence *= 0.5f;
    }

    // timers for persistence
    auto held_above = [&](float thresh, float hold_sec) -> bool {
      if (confidence >= thresh) {
        // if no previous high conf exists, set to now
        if (!this->_conf_above_since_set) {
          this->_conf_above_since = now;
          this->_conf_above_since_set = true;
        }
        // if previous high conf exists, calculate amount of time that high conf has been held
        // return true and transition to higher state
        return (now - this->_conf_above_since) >= hold_sec;
      } else {
        // if conf below threshold, reset the timer
        this->_conf_above_since_set = false;
        return false;
      }
    };

    auto held_below = [&](float thresh, float hold_sec) -> bool {
      if (confidence <= thresh) {
        if (!this->_conf_below_since_set) {
          this->_conf_below_since = now;
          this->_conf_below_since_set = true;
        }
        return (now - this->_conf_below_since) >= hold_sec;
      } else {
        this->_conf_below_since_set = false;
        return false;
      }
    };

    // state machine
    if (this->state == STATE_IDLE) {
      this->alerting = false;
      if (heat > 0.4f) {
        _transition(STATE_MONITOR, now);
      }

    } else if (this->state == STATE_MONITOR) {
      this->alerting = false;
      if (held_above(SUSPECT_ENTER_CONF, SUSPECT_HOLD_SEC)) {
        _transition(STATE_SUSPECT, now);
      } else if (heat < 0.2f && held_below(0.2f, 10.0f)) {
        _transition(STATE_IDLE, now);
      }

    } else if (this->state == STATE_SUSPECT) {
      this->alerting = false;
      if (held_above(CONFIRM_CONF, CONFIRM_HOLD_SEC)) {
        _transition(STATE_CONFIRMED, now);
      } else if (held_below(CLEAR_CONF, CLEAR_HOLD_SEC)) {
        _transition(STATE_MONITOR, now);
      }

    } else if (this->state == STATE_CONFIRMED) {
      // immediate to alerting
      this->alerting = true;
      _transition(STATE_ALERTING, now);

    } else if (this->state == STATE_ALERTING) {
      this->alerting = true;
      // clear conditions
      if (heat < 0.2f && held_below(CLEAR_CONF, CLEAR_HOLD_SEC)) {
        this->alerting = false;
        _transition(STATE_IDLE, now);
      }
    }

    Output out;
    out.state = state_name(this->state);
    out.heat_risk = heat;
    out.occupant_score = occ;
    out.confidence = confidence;
    out.thermal_ok = thermal_ok;
    out.co2_ok = co2_ok;
    out.motion_ok = motion_ok;
    out.alerting = this->alerting;
    return out;
  }

private:
  State state;
  float state_enter_ts;

  // initialize sensor readings
  ThermalReading thermal;
  CO2Reading co2;
  MotionReading motion;
  bool has_thermal;
  bool has_co2;
  bool has_motion;

  // initialize timestamps
  float last_motion_ts;
  float last_valid_thermal_ts;
  float last_valid_co2_ts;
  float last_valid_motion_ts;

  // initialize timestamp calculators
  bool _conf_above_since_set;
  bool _conf_below_since_set;
  float _conf_above_since;
  float _conf_below_since;

  // T/F is system in alert state?
  bool alerting;

  static float now_s() {
    return (float)millis() / 1000.0f;
  }

  float _heat_risk() {
    if (!this->has_thermal) { // if there is no thermal reading stored, return 0 heat risk
      return 0.0f;
    }
    // heat risk - weighted avg of t_avg + max
    float t_avg = this->thermal.t_avg;
    float t_max = this->thermal.t_max;
    // map 25C..45C to 0..1
    // change range -> change formulas
    // 25C = start of meaningful heat concern
    // 30C = starting reference point for when hotspots matter (moderate risk)
    // 45C = lower limit of extremely dangerous risk
    // divide by size of temp window to convert to dimensionless temp
    float risk_avg = clampf((t_avg - 25.0f) / 20.0f); // dimensionless temp
    float risk_max = clampf((t_max - 30.0f) / 20.0f);
    // weighted average of avg temp + max temp, can change weights if needed
    return clampf(0.6f * risk_avg + 0.4f * risk_max);
  }

  float _occupant_score(float now) {
    // reset occupant score
    float score = 0.0f;

    // thermal evidence
    if (this->has_thermal) {
      // reset thermal evidence score
      float thermal_evidence = 0.0f;
      // add all indicators of hotspot to thermal evidence score, with weights
      if (this->thermal.t_max >= this->HOTSPOT_HUMAN_MIN_C) {
        thermal_evidence += 0.5f;
      }
      thermal_evidence += 0.3f * clampf(this->thermal.hotspot_area);
      thermal_evidence += 0.2f * clampf(this->thermal.blob_score);
      // add thermal evidence score to total occupant score
      score += this->WT * clampf(thermal_evidence);
    }

    // CO2 evidence
    if (this->has_co2) {
      // reset CO2 evidence score
      float co2_evidence = 0.0f;
      // add all indicators of CO2 levels to CO2 score
      // weights do not add up to 1 because either can indicate risk
      if (this->co2.ppm >= this->CO2_HIGH_PPM) {
        co2_evidence += 0.7f;
      }
      if (this->co2.ppm_rate_per_min >= this->CO2_RISE_PPM_PER_MIN) {
        co2_evidence += 0.6f;
      }
      // add CO2 evidence score to total occupant score
      score += this->WC * clampf(co2_evidence);
    }

    // motion evidence
    // T/F is there recent motion? compare most recent motion time to upper limit
    bool motion_recent = (now - this->last_motion_ts) <= this->MOTION_RECENT_SEC;
    if (motion_recent) {
      // compute mconf level, currently can be only 0, 1, or direct confidence reading put into ingest_motion
      // check if mmWave module outputs detection confidence, signal strength, energy, or presence probability
      // if not, develop algo for detecting consecutive motion frames, reflection strength, small vs large motion, persistence of motion, filtering noise
      float mconf = (!this->has_motion) ? 1.0f : clampf(this->motion.confidence);
      score += this->WM * mconf;
    }

    // OBD evidence added to occupant score

    return clampf(score);
  }

  // T/F has the sensor reported data within the time limit for new readings?
  bool _sensor_ok(float last_ts, float now) {
    return (now - last_ts) <= this->SENSOR_TIMEOUT_SEC;
  }

  void _transition(State new_state, float now) {
    if (new_state != this->state) {
      // changes system state to current state
      this->state = new_state;
      // records when new state started
      this->state_enter_ts = now;
      // resets persistence timers
      this->_conf_above_since_set = false;
      this->_conf_below_since_set = false;
    }
  }
};

// Helper: compute avg/max + hotspot_area similarly to the Python class-side logic (done at ingest time here)
static void compute_thermal_derived(ThermalReading &r) {
  float sum = 0.0f;
  float mx = -1e9f;
  for (int i = 0; i < 8; i++) {
    for (int j = 0; j < 8; j++) {
      float t = r.temp_array[i][j];
      sum += t;
      if (t > mx) mx = t;
    }
  }
  r.t_avg = sum / 64.0f;
  r.t_max = mx;

  float hotspot_count = 0.0f;
  for (int i = 0; i < 8; i++) {
    for (int j = 0; j < 8; j++) {
      float t = r.temp_array[i][j];
      if (t >= r.t_max * 0.9f) { // tune this
        hotspot_count += 1.0f;
      }
    }
  }
  r.hotspot_area = hotspot_count / 64.0f;
}

static void compute_co2_derived(CO2Reading &r) {
  // ppm: float = 2000 * (th - 2) / (to - 4) # ppm concentration
  // NOTE: guard against divide-by-zero
  int denom = (r.to - 4);
  if (denom == 0) denom = 1;
  r.ppm = 2000.0f * (float)(r.th - 2) / (float)denom;
}

// example test with fake data
ChildHotCarDetector d;
float t0 = 0.0f;

void setup() {
  Serial.begin(9600); // it should be used with an arduino and serial rate 9600
  while (!Serial) { /* wait on some boards */ }

  t0 = (float)millis() / 1000.0f;
}

void loop() {
  static int i = 0;
  if (i >= 120) {
    delay(1000);
    return;
  }

  float now = t0 + (float)i;

  // fake thermal ramp
  ThermalReading tr;
  tr.ts = now;
  tr.blob_score = 0.6f;
  // Fill temp_array with something plausible; keep the ramp primarily in derived t_avg/t_max via assigned values below.
  for (int r = 0; r < 8; r++) {
    for (int c = 0; c < 8; c++) {
      tr.temp_array[r][c] = 30.0f + 0.05f * (float)i;
    }
  }
  // create a "hotspot" cell to push max upward
  tr.temp_array[3][3] = 33.0f + 0.08f * (float)i;

  compute_thermal_derived(tr);
  d.ingest_thermal(tr);

  // fake CO2 rising
  CO2Reading cr;
  cr.ts = now;
  cr.ppm = 800.0f + 6.0f * (float)i;
  cr.ppm_rate_per_min = 120.0f;
  d.ingest_co2(cr);

  // fake motion bursts
  MotionReading mr;
  mr.ts = now;
  mr.motion = (i % 15 == 0);
  mr.confidence = 0.9f;
  d.ingest_motion(mr);

  ChildHotCarDetector::Output out = d.update(now);
  if (i % 10 == 0) {
    Serial.print(i);
    Serial.print(" {");
    Serial.print("\"state\":\""); Serial.print(out.state); Serial.print("\",");
    Serial.print("\"heat_risk\":"); Serial.print(out.heat_risk, 4); Serial.print(",");
    Serial.print("\"occupant_score\":"); Serial.print(out.occupant_score, 4); Serial.print(",");
    Serial.print("\"confidence\":"); Serial.print(out.confidence, 4); Serial.print(",");
    Serial.print("\"thermal_ok\":"); Serial.print(out.thermal_ok ? "true" : "false"); Serial.print(",");
    Serial.print("\"co2_ok\":"); Serial.print(out.co2_ok ? "true" : "false"); Serial.print(",");
    Serial.print("\"motion_ok\":"); Serial.print(out.motion_ok ? "true" : "false"); Serial.print(",");
    Serial.print("\"alerting\":"); Serial.print(out.alerting ? "true" : "false");
    Serial.println("}");
  }

  i++;
  delay(1000);
}
